### DISCUSSION TITLE
[ Fundamentos da Orientação a Objetos - Using e Dispose] Quando devo usar o using + dispose?

### DISCUSSION THREAD

**Discussion by oimarcella** [balta-io/forum#403] [Tira Dúvidas] [answered]
Eu li outras duas dúvidas aqui no fórum sobre o uso do Dispose, mas ainda sim fiquei na dúvida em quais momentos eu deveria usar o using + dispose.

Na aula eu vi o exemplo com a classe Pagamento, e sua instância. Mas imagino que não deva ser em TODO o objeto que esse recurso deva ser utilizado, até porque temos o Garbage Collector.

Quero entender para saber se estou simplesmente adicionando mais linhas desnecessárias ao código ou não.

**Comment by andrebaltieri** [comment]
Bom dia, @mlasa, como vai?

A destruição dos objetos é algo que o coletor de lixo (Garbage Collector) do .NET já faz. O dispose é um método "especial" que podemos chamar quando esta ação acontece.

O using por sua vez, sinaliza que o objeto deve ser destruído logo após seu uso.

**Reply by oimarcella** [reply, marked as answer]
Mas como sei que posso deixar essa limpeza para o Garbage Collector fazer automaticamente, e quando eu deveria explicitamente destruir algum objeto? 
Por exemplo, eu sei que devo encerrar a conexão com banco para não deixar portas abertas e não comprometer limite de conexões.
Quais outros casos eu deveria saber?

**Reply by andrebaltieri** [reply]
@mlasa, ótima pergunta... a resposta é justamente essa que você colocou: A necessidade
Em caso de banco de dados e arquivos, por exemplo, onde não queremos correr o risco do coletor de lixo demorar para passar e manter nossos objetos em memória (Conexão ou arquivo abertos), aí usamos o dispose/using para já notificar o GC que o objeto pode ser destruído.

O GC é uma parte complexa, ele tem diferentes níveis de execução, a parte mais difícil para ele é justamente identificar quais objetos ele pode extinguir, por isto ajudamos ele com estas marcações.
